### ISSUE TITLE
deadline fails at date subtraction, not at the readout — day-count injection takes it 0.575 -> 0.963

### ISSUE THREAD

**Issue by 3x3xX3N0N** [jaredpalmer/kev#8]
# `deadline` fails at date subtraction, not at the readout — day-count injection takes it 0.575 → 0.963

PLAN.md concludes that the release-blocking `deadline` family is "a capability limit at <= 8B
for this readout, not a data gap", and ranks the fixes (1) an ordinal readout with cumulative
thresholds, (2) explicit day-count rendering, (3) capacity beyond 8B.

I ran a cheap inference-only check that suggests the ordering should be reversed: the readout is
fine, and the failure is specifically **date subtraction**.

## Measurement

80 unique `contrastive_deadline` items (`evals/v4/transfer-v4` development + test, deduped by
state+instructions), `jaredpalmer/kev-4b` in bf16 on one RTX 3080. No training, no changes to the
model — arm 1 only appends one sentence to `state["case"]` giving the day delta.

| arm | accuracy | predicts MIDDLE | expected-level MAE | Brier |
|---|---|---|---|---|
| kev-4b as shipped | 0.575 | 0.688 | 0.419 | 0.776 |
| + day count in the case text | **0.963** | 0.300 | 0.044 | 0.071 |

The middle level is 26.3% of labels, so 0.688 is severe hedging. Arm 0 reproduces the 0.45–0.60
reported in PLAN.md.

Confusion, arm 0 (rows = true, cols = predicted):

```
true 0 (on time):  [19, 21,  0]
true 1 (late):     [ 0, 21,  0]
true 2 (refused):  [ 0, 13,  6]
```

It collapses toward level 1 from both directions — 21/40 on-time and 13/19 refused items land in
the middle. Arm 1:

```
true 0: [37,  3,  0]
true 1: [ 0, 21,  0]
true 2: [ 0,  0, 19]
```

21/21 on the middle level, 19/19 on refused; all 3 remaining errors are on-time called late.

## Why this points at (2) rather than (1)

The injected sentence gives the **delta, not the answer**. The model still has to compare that
delta against a per-item threshold that varies across items (7 vs 14 days) and map the result to
three levels — and it does that 77/80 times with the unmodified head.

So the model can do numeric threshold comparison and the ordinal mapping; it cannot do date
subtraction. The middle-hedging looks like honest uncertainty rather than a readout pathology,
which would also explain why the two purpose-built training families didn't move it: they teach a
mapping that was never broken.

If that reading is right, an ordinal readout would mainly convert hedging into commitment. That
could raise the screen number while making calibration worse, which seems like the wrong trade for
this model.

## Reproduction

```python
# parse "Month D, YYYY" out of the deadline and received sentences, then:
delta = (received - deadline).days
state["case"] += f" The report arrived {delta} days after the deadline."   # or "before"
```

Load with `kev.evaluate.load("jaredpalmer/kev-4b", "cuda")` and `KEV_DTYPE=bf16`, then
`model.probs(model.encode(tok, rec))` per item.

## Caveats

- One family, 80 items, transfer set rather than the locked screen.
- Inference-time injection is an **upper bound** on what training with day-count rendering would
  give; it shows the ceiling is high, not that the training fix reaches it.
- I have not tried the ordinal readout, so this is evidence about (2), not a refutation of (1).

**Comment by 3x3xX3N0N** [comment]
very cool research dude keep up the good fight!!!

**Comment by jaredpalmer** [comment]
interesting! going to try this out!

**Comment by jaredpalmer** [comment]
Reproduced, and thank you — this reorders the plan.

**Your arm, same 40 development `deadline` items, `jaredpalmer/kev-4b` bf16 on an M5:**

| arm | acc | predicts middle | level MAE | Brier |
|---|---|---|---|---|
| as shipped | 0.525 | 0.750 | 0.452 | 0.829 |
| + "The report arrived N days after/before the deadline." | **0.925** | 0.350 | 0.077 | 0.139 |

On a Qwen3.5-4B-based checkpoint I trained today with the same recipe, the same injection takes it to **1.000** (Brier 0.001).

Independent evidence pointing the same way: the *untrained* Qwen3.5-9B base gets 0.82 on these items zero-shot (letter logits), Qwen3.5-4B 0.68 — and after our LoRA training, the adapted backbone read through the plain LM head drops to 0.53. So the model arrives able to subtract dates and our training data erodes it; the readout is not where the skill is lost. PLAN.md's ranking (ordinal readout first) was wrong, and I agree with your point that it would mostly convert honest hedging into commitment.

**One nuance.** Your sentence names the role ("after the *deadline*"). A *generic* annotator that lists every pairwise date difference as a `date_facts` field, without knowing which date is the deadline, helps much less: 0.525 → 0.550 (Kev-4B) and → 0.675 (Qwen3.5-4B checkpoint). So there are two gaps: the subtraction (large, yours) and binding a raw date fact to the policy's roles (smaller). A preprocessor that only finds dates gets the second-class result unless the model is trained to consume such facts.

**Plan, in order:** (1) render the date-bearing training families with relational day counts part of the time and with a generic `date_facts` field part of the time, so the model learns to use both — one 4B trial; your number is the ceiling; (2) then an opt-in `date_facts` request preprocessor; (3) ordinal readout last. Tracking in `PLAN_Qwen35.md`.

**Comment by jaredpalmer** [comment]
Closing: steps (1) and (2) of the plan above shipped, and the numbers landed where your ceiling said they would.

- **(1) training renderings.** The released `jaredpalmer/kev-9b`, `kev-4b` and `kev-0.8b` are deltas trained on the date-bearing families rendered with relational day counts part of the time and a generic `date_facts` field part of the time (`evals/night2/`, `scripts/build_night2_data.py`), so the model learns to consume both forms. The pre-delta checkpoints stay at Hub tag `v7-base`.
- **(2) preprocessor.** `KEV_DATE_FACTS=1` in `kev.serve` (and `--date_facts` in `kev.benchmark`) appends one sentence per pair of absolute dates found in the state ("June 26, 2026 is 8 days before July 4, 2026"): `kev.api.with_date_facts`. It is reported separately and never folded into the model's own numbers.
- **Result on `deadline`** (model cards): Kev-4B 0.60 → 0.85 with the preprocessor, Kev-9B 0.80 → 0.90 (Jev 0.93); overall out-of-domain accuracy for Kev-4B 0.797 → 0.820. That confirms your reading: the gap was date subtraction plus binding the fact to the policy's roles, not the readout.
- **(3) ordinal readout** stays deferred, for the reason you gave — it would mostly turn honest hedging into commitment. Tracked in `PLAN_Qwen35.md`.

Thanks again for the diagnosis; it reordered the plan.
